### ISSUE TITLE
output_verify="ignore" and "warn" option show unexpected behavior

### ISSUE THREAD

**Issue by ysBach** [astropy/astropy#7043]
From the [documentation](http://docs.astropy.org/en/stable/io/fits/usage/verification.html#verification-options), ``output_verify="ignore"`` or ``"warn"`` must not fix the non-standard FITS keywords when saving the FITS file using ``.writeto`` method, if I understood correctly.
 
But it worked somehow different from my expectation. First, the code I used for testing:

```python
import astropy
from astropy.io import fits
import numpy as np
print(astropy.__version__)
# 2.0.3
print(np.__version__)
# 1.13.3

'''
The test header looks like this:

NAXIS1  =                   10 / length of data axis 1                          
NAXIS2  =                   10 / length of data axis 2                          
TESTVERI= '2017/02/13-16:51:38 / Test VerifyWarning
'''

# code A (ignore)
testh = fits.Header.fromtextfile('test_hdr.txt')
testfits = fits.PrimaryHDU(data=np.ones((10,10)), header=testh)
testfits.writeto('test_ignoreA.fits', output_verify='ignore', overwrite=True)

# code A (warn)
testh = fits.Header.fromtextfile('test_hdr.txt')
testfits = fits.PrimaryHDU(data=np.ones((10,10)), header=testh)
testfits.writeto('test_warnA.fits', output_verify='warn', overwrite=True)

# code B
testh = fits.Header.fromtextfile('test_hdr.txt')
testfits = fits.PrimaryHDU(data=np.ones((10,10)), header=testh)
testfits.writeto('test_ignoreB.fits', output_verify='ignore', overwrite=True)
testfits.writeto('test_warnB.fits', output_verify='warn', overwrite=True)

# code C
testh = fits.Header.fromtextfile('test_hdr.txt')
testfits = fits.PrimaryHDU(data=np.ones((10,10)), header=testh)
testfits.writeto('test_warnC.fits', output_verify='warn', overwrite=True)
testfits.writeto('test_ignoreC.fits', output_verify='ignore', overwrite=True)
```

I only copied the header keyword ``TESTVERI`` here:
```
ignoreA.fits
TESTVERI= '''2017  '           / 02/13-16:51:38 / Test VerifyWarning            

ignoreB.fits
TESTVERI= '''2017  '           / 02/13-16:51:38 / Test VerifyWarning            

ignoreC.fits
TESTVERI= '2017/02/13-16:51:38 / Test VerifyWarning                             

warnA.fits
TESTVERI= '2017/02/13-16:51:38 / Test VerifyWarning                             

warnB.fits
TESTVERI= '''2017  '           / 02/13-16:51:38 / Test VerifyWarning            

warnC.fits
TESTVERI= '2017/02/13-16:51:38 / Test VerifyWarning                             
```
In other words, ``output_verify`` option for ``"ignore"`` and ``"warn"`` works correctly (i.e., do not change the keyword) if
1. only "warn" is used (using only "ignore" fixes the keyword)
2. "warn" is used first (if "ignore" is used first, even "warn" has fixed keyword)

This is quite different from my expectation, but I couldn't figure out why.

**Comment by dnfarias** [comment]
Hi! I'm trying to use the option output_verify = 'ignore' in flush()/close(), but it doesn't seem to work properly, because it gives me a lot of warnings and fixes! My images contains a lot of Non-Standard keywords (such as HIERARCH cards), that's why I'm posting it in this issue.

By the way, the astropy version is 2.0.2. I can attach an image example if it's necessary.

Thanks in advance

**Comment by dnfarias** [comment]
I'm going to detail this problem in a better way:

While using astropy.io on images with several HIERARCH cards (e.g., HIERARCH ESO INS RETA2 ROT = 0.0), which are Not-STANDARD FITS format, I encounter with a possible bug: if I do for example

hdulist = fits.open(IMAGE,mode='update')
hdulist[0].header.set('DISPAXIS', '1')
hdulist.close(output_verify = 'ignore')

It should prevent to print any warning message, but instead I get warning and ALSO fixes:

WARNING: VerifyWarning: Verification reported errors: [astropy.io.fits.verify]
WARNING: VerifyWarning: Card 'ESO INS IMAGE DISTOR1' is not FITS standard (invalid value string: '7.771690e-06 / distortion coefficient').  Fixed 'ESO INS IMAGE DISTOR1' card to meet the FITS standard. [astropy.io.fits.verify]
WARNING: VerifyWarning: Note: astropy.io.fits uses zero-based indexing.
 [astropy.io.fits.verify]
....

etc., a strange behavior that happened in the original post, so I think it could be a bug. I'm trying to trace where this happens, and the 'ignore' option it is used correctly once to verify in hdulist.py (astropy/io/fits/hdu) at 

def flush (self, output_verify='fix', verbose=False):
        self.verify(option=output_verify)

( def close uses flush so ..) and this is good, it does not print a thing. The bug appears in the last line of the same method
       .
       .
       .
   elif self._file.mode == 'update':
         self._flush_update() 

in self._flush_update(). If I search in this method, I could trace the 'bug' to 

    def _flush_update(self):
        """Implements flushing changes to a file in update mode."""

        for hdu in self:
             hdu._prewriteto(checksum=hdu._output_checksum, inplace=True)      

the hdu._prewriteto line. In this method, the warning appears after the return (end of the code)

    def _prewriteto(self, checksum=False, inplace=False):
        if self._scale_back:
            self._scale_internal(BITPIX2DTYPE[self._orig_bitpix],
                                 blank=self._orig_blank)
        self.update_header()
        if not inplace and self._data_needs_rescale:
            # Go ahead and load the scaled image data and update the header
            # with the correct post-rescaling headers
            _ = self.data

        return super(_ImageBaseHDU, self)._prewriteto(checksum, inplace)

but I don't know how to continue this task, so if you can help me it would be great. For now I'm gonna use 'warn' because it doesn't fix a thing.

**Comment by DFEvans** [comment]
The bug originates when the checksum of the HDUList is calculated. This includes converting each header card to a string, and when this occurs, the card calls self.verify('fix+warn') if it has not already been verified, at line 506 of astropy/io/fits/card.py

I'll leave it up to those more familiar with astropy to determine the best fix. Passing the "ignore" parameter would involve passing it through a large number of calls; should the object perhaps be marked as "verified" even if 'ignore' is used, to indicate that verify() has been called?

**Comment by saimn** [comment]
This may be the same as #5146, and probably also part of a longstanding issue with verification (#887), so the fix is not obvious.

@DFEvans - Thanks for investigating, the checksum computation may be a reason but I guess the bug is also present without this. The original issue or #5146 do not mention computing checksum, and I think the warning are always triggered by the conversion of the Header to a string when writing the file.
